pgjdbc/src/main/java/org/postgresql/jdbc/ResourceLock.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2004, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbc;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Extends a ReentrantLock for use in try-with-resources block.
+ *
+ * <h2>Example use</h2>
+ * <pre>{@code
+ *
+ *   try (ResourceLock ignore = lock.obtain()) {
+ *     // do something while holding the resource lock
+ *   }
+ *
+ * }</pre>
+ */
+public final class ResourceLock extends ReentrantLock implements AutoCloseable {
+
+  @SuppressWarnings("nullness")
+  private final Condition condition = newCondition();
```

Comment:
Do we need a condition for every `ResourceLock` instance?
Can we somehow make this lazy-initialized?

Reply:
We could remove it from ResourceLock and only create it on those types that need it. I think that would be much better than trying to make it "lazy-initialized" for the cases it's used (avoid racey lazy initialisation).

Reply:
The last commit removes condition from `ResourceLock` and instead it's used explicitly where needed. See https://github.com/pgjdbc/pgjdbc/pull/2635/commits/a66af8819cf152324d42969bb08431123ca0b9cd?diff=unified&w=1

Reply:
We have it in a single place only, so that is fine to just have a field there.